cores/esp8266/core_esp8266_si2c.c

```diff
@@ -211,4 +312,575 @@ uint8_t twi_status(){
 
     if(!twi_write_start()) return I2C_SDA_HELD_LOW_AFTER_INIT;  //line busy. SDA again held low by another device. 2nd master?
     else                   return I2C_OK;       				//all ok 
-}
\ No newline at end of file
+}
+
+uint8_t twi_transmit(const uint8_t* data, uint8_t length)
+{
+  uint8_t i;
+
+  // ensure data will fit into buffer
+  if (length > TWI_BUFFER_LENGTH) {
+    return 1;
+  }
+  
+  // ensure we are currently a slave transmitter
+  if (twi_state != TWI_STX) {
+    return 2;
+  }
+  
+  // set length and copy data into tx buffer
+  twi_txBufferLength = length;
+  for (i = 0; i < length; ++i) {
+    twi_txBuffer[i] = data[i];
+  }
+  
+  return 0;
+}
+
+void twi_attachSlaveRxEvent( void (*function)(uint8_t*, int) )
+{
+  twi_onSlaveReceive = function;
+}
+
+void twi_attachSlaveTxEvent( void (*function)(void) )
+{
+  twi_onSlaveTransmit = function;
+}
+
+void ICACHE_RAM_ATTR twi_reply(uint8_t ack)
+{
+  // transmit master read ready signal, with or without ack
+  if (ack) {
+    //TWCR = _BV(TWEN) | _BV(TWIE) | _BV(TWINT) | _BV(TWEA);
+	SCL_HIGH();		// _BV(TWINT)
+	twi_ack = 1; 	// _BV(TWEA)
+  } else {
+	//TWCR = _BV(TWEN) | _BV(TWIE) | _BV(TWINT);
+	SCL_HIGH();		// _BV(TWINT)
+	twi_ack = 0;	// ~_BV(TWEA)
+  }
+}
+
+#if 1
+void ICACHE_RAM_ATTR twi_stop(void)
+{
+	// send stop condition
```

Comment:
The reset of the file is indented with spaces, please make indentation consistent.